File: swift/510_InorderSuccessorInBSTII.swift
class Solution {
    func inorderSuccessor(_ candidate: Node?) -> Node? {
        var node: Node? = candidate
        if node?.right != nil {
            node = node?.right
            while node?.left != nil { node = node?.left }
            return node
        }
        
        while node?.parent != nil && node?.parent?.right === node { node = node?.parent }
        return node?.parent
    }
}

File: cpp/510_InorderSuccessorInBSTII.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* parent;
};
*/

class Solution {
public:
    Node* inorderSuccessor(Node* node) {
        if (!node) return NULL;

        int val = node->val;
        Node* curr = node->parent;
        Node* cand = NULL;
        while (curr) {
            if (curr->val > val) {
                cand = curr;
                break;
            }
            curr = curr->parent;
        }
        
        curr = node->right;
        while (curr) {
            if (curr->val > val && (cand == NULL || curr->val < cand->val)) cand = curr;
            curr = curr->left;
        }
        return cand;
    }
};
